document/js-api/index.bs

```diff
@@ -977,6 +980,7 @@ This slot holds a [=function address=] relative to the [=surrounding agent=]'s [
     1. Let (|store|, |ret|) be the result of [=func_invoke=](|store|, |funcaddr|, |argsSeq|).
     1. Set the [=surrounding agent=]'s [=associated store=] to |store|.
     1. If |ret| is [=error=], throw an exception. This exception should be a WebAssembly {{RuntimeError}} exception, unless otherwise indicated by <a href="#errors">the WebAssembly error mapping</a>.
+    1. If |ret| is an exception, throw a {{RuntimeError}} exception.
```

Comment:
Currently the implementation just rethrows the existing exception (i.e. the "anyref" representing the exception package in WebAssembly), which happens to be an instance of {{RuntimeError}}. I totally see need to make these distinguishable from a {{RuntimeError}} thrown for the "error" case on line above though.

Reply:
How does that work? The core spec can't create RuntimeErrors, AIUI. Also, I get RuntimeErrors even when I'm throwing an anyref value I passed in from JS (https://github.com/web-platform-tests/wpt/pull/18294).

Reply:
I see your point of having to keep out the {{RuntimeError}} out of the core spec. Agreed.

However there might be situations where the exception being propagated to the embedding (i.e. JavaScript) is not a {{RuntimeError}}. I am thinking about the case where the function just uses rethrow on an existing exnref. In this case the exception package being propagated to the JavaScript would be the original exception, not necessarily a {{RuntimeError}}. At least that was my current understanding, but I am open to specifying different semantics of course. Here is the sketched web-platform test I have in mind (just for illustration, might contain typos):

```
promise_test(async () => {
  const kWasmAnyRef = 0x6f;
  const kSig_v_r = makeSig([kWasmAnyRef], []);
  const kSig_r_v = makeSig([], [kWasmAnyRef]);
  const builder = new WasmModuleBuilder();
  const fnIndex = builder.addImport("module", "fn", kSig_v_v);
  const except = builder.addException(kSig_v_r);
  builder.addFunction("catch_and_rethrow", kSig_r_v)
    .addBody([
      kExprTry, kWasmStmt,
        kExprCallFunction, fnIndex,
      kExprCatch,
        kExprRethrow,
      kExprEnd,
      kExprRefNull,
    ])
    .exportFunc();

  const buffer = builder.toBuffer();

  class MyError extends Error {};
  const error = new MyError();
  const fn = () => { throw error };
  const {instance} = await WebAssembly.instantiate(buffer, {
    module: { fn }
  });
  assert_throws(new MyError(), () => instance.exports.catch_and_rethrow());
}, "Imported JS function throws, Wasm catches and rethrows");
```

Reply:
Okay, this makes sense to me. What doesn't make sense is how `kExprRefNull` gets turned into a RuntimeError in a case like this:

```js
promise_test(async () => {
  const kWasmAnyRef = 0x6f;
  const kSig_v_r = makeSig([kWasmAnyRef], []);
  const builder = new WasmModuleBuilder();
  const except = builder.addException(kSig_v_r);
  builder.addFunction("throw_null", kSig_v_v)
    .addBody([
      kExprRefNull,
      kExprThrow, except,
    ])
    .exportFunc();
  const buffer = builder.toBuffer();
  const {instance} = await WebAssembly.instantiate(buffer, {});
  assert_throws(new WebAssembly.RuntimeError(), () => instance.exports.throw_null());
}, "Wasm function throws null");
```

What is special about getting an anyref value into Wasm by throwing it in JS and catching it in Wasm, as opposed to creating it in Wasm or passing it as an argument?

Reply:
I don't think there is anything special about getting an `anyref` value passed into Wasm this way. The difference is between `kExprThrow` and `kExprRethrow` I feel. The model I am currently imagining is the following:

- `kExprThrow`: This will always create a new exception package, wrapping values according to the exception signature, as well as some internal notion of stack trace (or means of reconstructing it) into a package. This package has type `exnref`. Should such a package hit JavaScript (e.g. via stack unwinding) then it should be represented as an instance of RuntimeError (or some other, yet to be determined error constructor).

- `kExprRethrow`: This will take an existing exception package of type `exnref` and use it as is. Such an exception package might be one that was created by Wasm (e.g. by `kExprThrow` as described above). It could however be any other JavaScript value that was passed into Wasm as well. The `exnref` type is currently inhabited by the same set of values as the `anyref` type. Should an exception package raised via `kExprRethrow` hit JavaScript, then it is _not_ guaranteed to be an instance of RuntimeError, it can be any value (the same as the originally passed value).

Note that all of the above is just describing the model that V8 follows at the moment. If the JavaScript API ends up spec'ing this differently, then V8 will of course adapt according to the spec.

For example I could very well imagine a model where `exnref` type is only inhabited by instances of RuntimeError. This would required an arbitrary JavaScript value caught by Wasm to be implicitly wrapped into an exception package. Should Wasm rethrow this package via `kExprRethrow`, then it would be represented as an instance of RuntimeError (not the same as the originally thrown value) to JavaScript. Would such a model make more sense?